cluster-api/machines-api/proposal.md

```diff
@@ -0,0 +1,164 @@
+replacing it, most tools would opt to use this pattern to attempt an in-place
+modification first, before falling back to a full replacement.
+
+It seems like a much more powerful concept to allow every tool to instead say:
+
+    for machine in machines:
+        update machine
+
+and allow the provider to decide if it is capable of performing an in-place
+update, or if a full Node replacement is necessary.
+
+## Omitted Capabilities
+
+* A scalable representation of a group of nodes
+
+Given the existing targeted capabilities, this functionality could easily be
+built client-side via label selectors to find groups of Nodes and using (1) and
+(2) to add or delete instances to simulate this scaling.
+
+It is natural to extend this API in the future to introduce the concepts of
+MachineSets and MachineDeployments that mirror ReplicaSets and Deployments, but
+an initial goal is to first solidify the definition and behavior of a single
+Machine, similar to how Kubernetes first solidifed Pods.
+
+A nice property of this proposal is that if provider controllers are written
+solely against Machines, the concept of MachineSets can be implemented in a
+provider-agnostic way with a generic controller that uses the MachineSet
+template to create and delete Machine instances. All Machine-based provider
+controllers will continue to work, and will get full MachineSet functionality
+for free without modification. Similarly, a MachineDeployment controller could
+then be introduced to generically operate on MachineSets without having to know
+about Machines or providers. Provider-specific controllers that are actually
+responsible for creating and deleting hosts would only ever have to worry about
+individual Machine objects, unless they explicitly opt into watching
+higher-level APIs like MachineSets in order to take advantage of
+provider-specific features like AutoScalingGroups or Managed Instance Groups.
+
+However, this leaves the barrier to entry very low for adding new providers:
+simply implement creation and deletion of individual Nodes, and get Sets and
+Deployments for free.
+
+* A provider-agnostic mechanism to request new nodes
+
+In this proposal, only certain attributes of Machines are provider-agnostic and
+can be operated on in a generic way. In other iterations of similar proposals,
+much care had been taken to allow the creation of truly provider-agnostic
+Machines that could be mapped to provider-specific attributes in order to better
+support usecases around automated Machine scaling. This introduced a lot of
+upfront complexity in the API proposals.
+
+This proposal starts much more minimalistic, but doesn't preclude the option of
+extending the API to support these advanced concepts in the future.
+
+* Dynamic API endpoint
+
+This proposal lacks the ability to declaratively update the kube-apiserver
+endpoint for the kubelet to register with. This feature could be added later,
```

Comment:
I'm unclear on what this section means, but it kinda sounds like something CRD could handle?